### PR TITLE
STAT: Apply UCX statistics filtering to UDP report

### PR DESCRIPTION
**Pull request by shuki-zanyovka**
## What
This update applies statistics filtering to the UDP report.

## Why ?
Statistics filtering is already available on UCX but not applied to the UDP server report (only to stdout report).

## How ?
By applying it using the current UCX statistics SW architecture.

### CODE REVIEW

**Comment by brminich** [comment]
ok to test

File: src/ucs/stats/serialization.c [hunk 1]
```diff
@@ -283,11 +311,17 @@ ucs_stats_serialize_binary(FILE *stream, ucs_stats_node_t *root,
...
         cls = elem->cls;
         ucs_stats_write_str(cls->name, stream);
-        FWRITE_ONE(&cls->num_counters, stream);
-        for (counter = 0; counter < cls->num_counters; ++counter) {
-            ucs_stats_write_str(cls->counter_names[counter], stream);
+        FWRITE_ONE(&elem->filter_node->num_counters_filtered, stream);
+
+        filter_node = elem->filter_node;
+        for (counter = 0; (counter < cls->num_counters) && (counter < 64); ++counter) {
+            if (filter_node->counters_bitmask & UCS_BIT(counter)) {
```
Comment (yosefe) [inline src/ucs/stats/serialization.c:322]: ucs_for_each_bit

File: src/ucs/stats/serialization.c [hunk 2]
```diff
@@ -244,8 +250,30 @@ ucs_stats_serialize_binary_recurs(FILE *stream, ucs_stats_node_t *node,
...
+    /* 
+       Fix warning: uninit_use_in_call: Using uninitialized element of 
+                    array "filtered_counters" when calling "ucs_stats_write_counters"
+    */
+    filtered_counters[0] = 0; 
+
+    filter_node = node->filter_node;
+    j = 0;   
+    for (i = 0; (i < node->cls->num_counters) && (i < 64); ++i) {
+        if (filter_node->counters_bitmask & UCS_BIT(i)) {
```
Comment (yosefe) [inline src/ucs/stats/serialization.c:262]: ucs_for_each_bit

File: src/ucs/stats/serialization.c [hunk 3]
```diff
@@ -244,8 +250,30 @@ ucs_stats_serialize_binary_recurs(FILE *stream, ucs_stats_node_t *node,
...
+            ucs_stats_counter_t counters_acc = 0;
```
Comment (yosefe) [inline src/ucs/stats/serialization.c:263]: maybe just declare
 ucs_stats_counter_t filtered_counters[64] ={}
and then
  filtered_counters[j] += temp_node->counters[i];

File: src/ucs/stats/serialization.c [hunk 4]
```diff
@@ -232,6 +234,10 @@ ucs_stats_serialize_binary_recurs(FILE *stream, ucs_stats_node_t *node,
     ucs_stats_clsid_t *elem, search;
     ucs_stats_node_t *child;
     uint8_t sentinel;
+    ucs_stats_filter_node_t *filter_node;
+    ucs_stats_counter_t filtered_counters[64];
+    uint32_t i;
+    uint32_t j;
```
Comment (yosefe) [inline src/ucs/stats/serialization.c:240]: can we have more meaningful names?

File: src/ucs/stats/stats.c [hunk 5]
```diff
@@ -354,15 +355,19 @@ static void ucs_stats_add_to_filter(ucs_stats_node_t *node,
...
         filter_index = ucs_config_names_search(ucs_global_opts.stats_filter,
                                                node->cls->counter_names[i]);
         if (filter_index >= 0) {
             filter_node->counters_bitmask |= UCS_BIT(i);
+            num_counters_out++;
             found = 1; 
         }
     }
 
+    filter_node->num_counters_filtered = num_counters_out;
```
Comment (yosefe) [inline src/ucs/stats/stats.c:369]: why need this field?
isn't it always popcount of counters_bitmask ?

File: src/ucs/stats/serialization.c (hunk 1 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:322]: Fixed.

File: src/ucs/stats/serialization.c (hunk 2 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:262]: Fixed.

File: src/ucs/stats/serialization.c (hunk 3 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:263]: Fixed.

File: src/ucs/stats/serialization.c (hunk 4 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:240]: Fixed.

File: src/ucs/stats/stats.c (hunk 5 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/stats.c:369]: Hi @yosefe , 
This field is required since we need to write it before iterating through the bitmap so it makes sense to store it beforehand,
ucs_stats_serialize_binary() ==>  FWRITE_ONE(&elem->filter_node->num_counters_filtered, stream);

Comment (yosefe) [reply src/ucs/stats/stats.c:369]: why not compute popcount(node->counters_bitmask) and write it?

Comment (shuki-zanyovka) [reply src/ucs/stats/stats.c:369]: OK, I see - Sounds good, fixed.

File: src/ucs/stats/libstats.h [hunk 6]
```diff
@@ -132,12 +132,12 @@ struct ucs_stats_node {
...
+    ucs_list_link_t           list;                  /* nodes sharing same parent.*/
     ucs_list_link_t           children;
-    ucs_list_link_t           type_list_head;     /* nodes with same ancestors classes */
-    int                       type_list_len;      /* length of list */
-    int                       ref_count;          /* report node when non zero */
-    uint64_t                  counters_bitmask;   /* which counters to print */
+    ucs_list_link_t           type_list_head;        /* nodes with same ancestors classes */
+    int                       type_list_len;         /* length of list */
+    int                       ref_count;             /* report node when non zero */
+    uint64_t                  counters_bitmask;      /* which counters to print */
```
Comment (brminich) [inline src/ucs/stats/libstats.h:140]: plz revert this change

File: src/ucs/stats/serialization.c [hunk 7]
```diff
@@ -232,6 +234,10 @@ ucs_stats_serialize_binary_recurs(FILE *stream, ucs_stats_node_t *node,
     ucs_stats_clsid_t *elem, search;
     ucs_stats_node_t *child;
     uint8_t sentinel;
+    ucs_stats_filter_node_t *filter_node;
+    ucs_stats_counter_t filtered_counters[64] = {};
```
Comment (brminich) [inline src/ucs/stats/serialization.c:238]: plz, move initialized field to the very beginning of the func

File: src/ucs/stats/serialization.c [hunk 8]
```diff
@@ -244,8 +250,22 @@ ucs_stats_serialize_binary_recurs(FILE *stream, ucs_stats_node_t *node,
...
+    filter_node = node->filter_node;
+    filtered_counter_index = 0;
+    ucs_for_each_bit(counter_index, filter_node->counters_bitmask) {
+        ucs_stats_node_t *temp_node;
+
+        ucs_list_for_each(temp_node, &filter_node->type_list_head, type_list) {
+            filtered_counters[filtered_counter_index] += temp_node->counters[counter_index];
+        }
+
+        /* Counters */
```
Comment (brminich) [inline src/ucs/stats/serialization.c:263]: what does this comment mean?

File: src/ucs/stats/serialization.c [hunk 9]
```diff
@@ -283,11 +303,18 @@ ucs_stats_serialize_binary(FILE *stream, ucs_stats_node_t *root,
...
+        ucs_stats_filter_node_t *filter_node;
+        unsigned num_counters_filtered;
+
         cls = elem->cls;
         ucs_stats_write_str(cls->name, stream);
-        FWRITE_ONE(&cls->num_counters, stream);
-        for (counter = 0; counter < cls->num_counters; ++counter) {
-            ucs_stats_write_str(cls->counter_names[counter], stream);
+
+        filter_node = elem->filter_node;
```
Comment (brminich) [inline src/ucs/stats/serialization.c:312]: minor: please align by `=` with the line below

File: src/ucs/stats/stats.c [hunk 10]
```diff
@@ -327,6 +327,7 @@ static void ucs_stats_add_to_filter(ucs_stats_node_t *node,
     int found = 0;
     int filter_index = 0;
     int i;
+    unsigned num_counters_out;
```
Comment (brminich) [inline src/ucs/stats/stats.c:330]: why do you need this variable?

File: src/ucs/stats/libstats.h (hunk 6 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/libstats.h:140]: Fixed.

File: src/ucs/stats/serialization.c (hunk 8 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:263]: Removed it.

File: src/ucs/stats/serialization.c (hunk 9 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:312]: Do you mean like this:
        filter_node                   = elem->filter_node;
        num_counters_filtered = __builtin_popcount(elem->filter_node->counters_bitmask);

?

Fixed.

File: src/ucs/stats/stats.c (hunk 10 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/stats.c:330]: True, it's not needed anymore - Removed it.

File: src/ucs/stats/serialization.c (hunk 7 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:238]: Fixed.

File: src/ucs/stats/serialization.c [hunk 11]
```diff
@@ -244,8 +251,20 @@ ucs_stats_serialize_binary_recurs(FILE *stream, ucs_stats_node_t *node,
     /* Name */
     ucs_stats_write_str(node->name, stream);
 
-    /* Counters */
-    ucs_stats_write_counters(node->counters, cls->num_counters, stream);
+    /* Filter output */
+    filter_node = node->filter_node;
+    filtered_counter_index = 0;
+    ucs_for_each_bit(counter_index, filter_node->counters_bitmask) {
+        ucs_stats_node_t *temp_node;
```
Comment (yosefe) [inline src/ucs/stats/serialization.c:258]: can move this to line 239

File: src/ucs/stats/serialization.c [hunk 12]
```diff
@@ -244,8 +251,20 @@ ucs_stats_serialize_binary_recurs(FILE *stream, ucs_stats_node_t *node,
     /* Name */
     ucs_stats_write_str(node->name, stream);
 
-    /* Counters */
-    ucs_stats_write_counters(node->counters, cls->num_counters, stream);
+    /* Filter output */
+    filter_node = node->filter_node;
+    filtered_counter_index = 0;
```
Comment (yosefe) [inline src/ucs/stats/serialization.c:256]: align = on column

File: src/ucs/stats/serialization.c [hunk 13]
```diff
@@ -283,11 +302,18 @@ ucs_stats_serialize_binary(FILE *stream, ucs_stats_node_t *root,
...
+        unsigned num_counters_filtered;
+
         cls = elem->cls;
         ucs_stats_write_str(cls->name, stream);
-        FWRITE_ONE(&cls->num_counters, stream);
-        for (counter = 0; counter < cls->num_counters; ++counter) {
-            ucs_stats_write_str(cls->counter_names[counter], stream);
+
+        filter_node           = elem->filter_node;
+        num_counters_filtered = __builtin_popcount(elem->filter_node->counters_bitmask);
```
Comment (yosefe) [inline src/ucs/stats/serialization.c:312]: ucs_popcount

File: src/ucs/stats/serialization.c (hunk 11 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:258]: Fixed.

File: src/ucs/stats/serialization.c (hunk 13 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:312]: Fixed.

File: src/ucs/stats/serialization.c (hunk 12 above)
Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c:256]: Fixed.

File: src/ucs/stats/serialization.c (hunk 11 above)
Comment (yosefe) [reply src/ucs/stats/serialization.c:256]: pls move " ucs_stats_node_t *temp_node;" to line 239

Comment (shuki-zanyovka) [reply src/ucs/stats/serialization.c]: Fixed.

**Review by yosefe — APPROVED** [review]
pls squash

**Review by brminich — APPROVED** [review]

**Comment by shuki-zanyovka** [comment]
Hi @yosefe ,

I can't access the log of the failure (last test seems to not have results or so) - I suspect we just need to re-run the test - How can I find out the reason for the failure?

Regards,
Shuki

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
VM issue
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest
